towhee/hub/file_manager.py

```diff
@@ -0,0 +1,393 @@
+# Copyright 2021 Zilliz. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+from pathlib import Path
+from typing import Union, List
+from shutil import copy2, copytree, rmtree
+
+
+from towhee.engine.singleton import singleton
+from towhee.engine import DEFAULT_LOCAL_CACHE_ROOT
+from towhee.hub.hub_tools import download_repo
+
+@singleton
+class FileManagerConfig():
+    """
+    Global FileManager config
+
+    This class dictates which cache locations to look through and in what order to check
+    for files. This config is ultimately used with FileManager.
+
+    Args:
+        set_default_cache (`str` | `Path`):
+            The default cache to check in, if nothing supplied, the default cache of $HOME/.towhee
+            will be used.
+    """
+
+    def __init__(self):
+        # TODO: #1 Deal with specifying cache priority per pipeline?
+        self._default_cache = DEFAULT_LOCAL_CACHE_ROOT
+        self._cache_paths = [DEFAULT_LOCAL_CACHE_ROOT]
+        self._cache_operators = []
+        self._cache_pipelines = []
+
+        self._default_lock = threading.Lock()
+        self._paths_lock = threading.Lock()
+        self._operators_lock = threading.Lock()
+        self._pipelines_lock = threading.Lock()
+
+    @property
+    def default_cache(self):
+        return self._default_cache
+
+    @property
+    def cache_paths(self):
+        return self._cache_paths
+
+    @property
+    def cache_operators(self):
+        return self._cache_operators
+
+    @property
+    def cache_pipelines(self):
+        return self._cache_pipelines
+
+    @property
+    def operator_lock(self):
+        return self._operators_lock
+
+    @property
+    def pipeline_lock(self):
+        return self._pipelines_lock
+
+    def change_default_cache(self, default_path: Union[str, Path]):
+        default_path = Path(default_path)
+        with self._paths_lock and self._default_lock:
+            if not self._cache_paths:
+                self._cache_paths.append(default_path)
+            self._cache_paths[-1] = default_path
```

Comment:
Yea, the last path is always the default cache, except if someone purposefully deletes it then runs this. Im thinking of hiding delete_cache for now because it is kinda of pointless.

Reply:
Yes, the last path should always be the current default path, so we overwrite it when updating. I made some changes just in case someone removes the default path before doing this step.